exercises/spectre/descr.md

```diff
@@ -0,0 +1,299 @@
+
+In fact, these two lists are sufficient to describe this tree in an
+unambiguous manner: no other three has the same list of leaf data and
+the same list of leaf depths.
+
+Let us combine these lists, and use the word *spectre* to refer a list
+of pairs of a datum and a depth:
+
+```
+  type depth =
+    int
+  type 'a spectre =
+    ('a * depth) list
+```
+
+The spectre of the tree
+`Fork (Leaf 1, Fork (Leaf 2, Leaf 3))`
+is
+`[(1, 1); (2, 2); (3, 2)]`.
+
+A tree is unambiguously described by its spectre: two distinct trees must have
+distinct spectres.
+
+This raises the question: from a spectre, how does one reconstruct a tree?
+
+This simple question gives rise to a nontrivial exercise,
+which can be solved in several ways.
+
+This is really a *parsing* problem. Indeed, parsing consists in reconstructing
+a tree out of a purely sequential representation of this tree. Furthermore,
+because the spectre of a tree carries the depth of each leaf, reconstructing
+a tree from a spectre is analogous to *indentation-aware parsing*.
+
+## Warmup: From a Tree to its Spectre
+
+**Question 1.** Define a function `elements` of type `'a tree -> 'a list`
+that computes the list of the data values carried by the leaves of a tree.
+
+**Question 2.** Define a function `depths` of type `'a tree -> depth list`
+that computes the list of the depths of the leaves of a tree.
+
+**Question 3.** Define a function `spectre` of type `'a tree -> 'a spectre`
+that computes the spectre of a tree.
+
+*Note.* The above three questions have naïve solutions that use the list
+concatenation operation `@` and (because of that) have quadratic time
+complexity. It is possible to avoid the use of concatenation by defining
+auxiliary functions that take an accumulator (a list) as an extra argument.
+This yields solutions that have linear time complexity.
+
+## Setting up an Input Channel
+
+In preparation for the next question, let us construct a simple notion of
+*input channel*.
+
+An input channel helps us read a sequence of things: for instance, a sequence
+of bits, a sequence of characters, or (in this exercise) a spectre, that is, a
+sequence of pairs of a datum and a depth.
+
+Although we could work with an explicit representation of the input (say, as a
+list of things that remain to be read), it is preferable to adopt a more
+abstract point of view and to view an *input channel* as an abstract object.
+It is useful to think of an input channel as a *read head* that moves along
+a fixed *tape*. An input channel supports three operations, namely:
+
+* `peek()` returns `Some x`, where `x` is the element in front of the read
+  head, if there is one. It returns `None` if the input has been exhausted,
+  that is, if the read head has reached the end of the input.
+
+* `consume()` discards the next input element. In other words, it moves the
+  read head forward by one element. Invoking `consume` is permitted only if
+  the read head has *not* yet reached the end of the input.
```

Comment:
...and what happens if `consume` is called while the head reached EOF?

Reply:
This is unspecified. The student need not worry about this case; the automated grader will be careful to never call `consume` in a situation where the head has reached the end of the input.